include/core/CProgramCounters.h

```diff
@@ -0,0 +1,224 @@
+    E_TSADNumberRecordsNoTimeField,
+
+    //! The number of errors converting time fields to numeric values
+    E_TSADNumberTimeFieldConversionErrors,
+
+    //! The number of records with out-of-order time fields
+    E_TSADNumberTimeOrderErrors,
+
+    //! The number of new attributes created in the data gatherer
+    E_TSADNumberNewAttributes,
+
+    //! The number of new attributes not created in the data gatherer
+    //! because there wasn't enough free resource
+    E_TSADNumberNewAttributesNotAllowed,
+
+    //! The number of new people put into a recycled slot
+    E_TSADNumberNewAttributesRecycled,
+
+    //! The number of 'over' fields created
+    E_TSADNumberOverFields,
+
+    //! The number of 'by' fields created
+    E_TSADNumberByFields,
+
+    //! The number of times "ExcludeFrequent" has been invoked by the model
+    E_TSADNumberExcludedFrequentInvocations,
+
+    //! The number of samples received outside the latency window
+    E_TSADNumberSamplesOutsideLatencyWindow,
+
+    //! The number of model creation failures from being over memory limit
+    E_TSADNumberMemoryLimitModelCreationFailures,
+
+    //! The number of old people or attributes pruned from the models
+    E_TSADNumberPrunedItems,
+
+    //! The number of times partial memory estimates have been carried out
+    E_TSADNumberMemoryUsageEstimates,
+
+    // Add any new values here
+
+    //! This MUST be last
+    E_LastEnumCounter
+};
+
+using TCounterTypeSet = std::set<ECounterTypes>;
+}
+
+namespace core {
+
+class CStateRestoreTraverser;
+class CStatePersistInserter;
+
+struct SCounterDefinition {
+    ml::counter_t::ECounterTypes s_Type;
+    std::string s_Name;
+    std::string s_Description;
+};
+
+//! \brief
+//! A collection of runtime global counters
+//!
+//! DESCRIPTION:\n
+//! A static collection of runtime global counters, atomically incremented
+//!
+//! To add a new counter, add it to the counter_t::ECounterTypes enum in
+//! the penultimate position and add its details and description to the
+//! m_CounterDefinitions array.
+//! In addition to this, individual programs should add the new enum type to
+//! the collection passed to registerProgramCounterTypes.
+//!
+//! Enum names should have a meaningful prefix to effectively namespace counters
+//! used in separate applications, e.g. \c TSAD is used for the Time Series Anomaly Detection
+//! counters.
+//!
+//! IMPLEMENTATION DECISIONS:\n
+//! A singleton class: there should only be one collection of global counters
+//!
+class CORE_EXPORT CProgramCounters : private CNonCopyable {
+private:
+    using TCounter = std::atomic_uint_fast64_t;
+    using TCounterArray = std::array<TCounter, counter_t::E_LastEnumCounter>;
+    using TCounterDefinitionArray = std::array<SCounterDefinition, counter_t::E_LastEnumCounter>;
+    using TUInt64Vec = std::vector<uint64_t>;
+
+public:
+    //! Singleton pattern
+    static CProgramCounters& instance();
```

Comment:
The `m_Counters` member is not static, so the class does need to be a singleton rather than a non-instantiatable grouping of all `static` functionality.

I think the lack of usage of `instance()` outside of the class itself is to reduce verbosity when incrementing the counters, so I would be in favour of dropping this method.  But then it's weird to have a singleton without an `instance()` method, so there should be a comment to say that it's to reduce verbosity.